### ISSUE TITLE
Using esri-loader w/ StealJS

### ISSUE THREAD

**Issue by tomwayson** [roemhildtg/can-arcgis#10]
Hey @roemhildtg 

This is a very interesting project. I was a fan of CanJS back in the day and had a dream of getting it to work w/ the ArcGIS API for JavaScript, but I don't think I ever did, so kudos to you.

Esri-loader has [recenlty added a promise-based API based largely on the work in esri-promise](https://github.com/Esri/esri-loader/releases/tag/v2.0.0). It looks like you are using esri-promise now, but maybe at one point you [may have used esri-loader](https://github.com/roemhildtg/can-arcgis/blob/920f0c8b26db6a2ce7d1c45470eb6f39d90b7ecb/steal-dojo.js#L2)? It looks like you were trying to maybe set up a steal plugin to allow you to use `import` statements for esri classes in the widgets? I'm guessing that b/c I see you [exclude esri modules from the steal build](https://github.com/roemhildtg/can-arcgis/blob/727f3b2bac4a3c433a8404cf2775e14fb4f90d5a/build.js#L12-L14) but I don't see why you need that now. I'm curious b/c I was considering setting up a StealJS example for esri-loader, but it looks like you've already been down some interesting paths and I wonder what the result was.

It looks like you may have shifted to a new paradigm, where your widget configurations include an `optionsPromise`. You could do this now w/ esri-loader too. For example:

```js
// viewer.js config file
import { loadModules } from 'esri-loader';

export default {
    // ...
    // other config properties
    // ...
    widgets: [{
    path: 'esri/widgets/BasemapToggle',
    parent: 'view',
    type: 'esri',
    position: 'top-left',
    iconClass: 'esri-icon-basemap',
    optionsPromise: loadModules([
            'esri/Basemap', 
            'esri/layers/TileLayer',
            'esri/layers/MapImageLayer'
        ]).then(([Basemap, TileLayer, MapImageLayer]) => {
            const base = new Basemap({
                thumbnailUrl: 'https://js.arcgis.com/4.5/esri/images/basemap/hybrid.jpg',
                id: 'aerial',
                baseLayers: [new TileLayer({
                    url: `${url}/basemaps/imagery_2016/MapServer`
                })],
                referenceLayers: [new MapImageLayer({
                    url: `${url}/basemaps/gray_labels/MapServer`
                })]
            });
            return {
                nextBasemap: base
            };
        });
  },
}
```

Also, it looks like you are [using your own fork of esri-promise to avoid having to use yet another promise implementation](https://github.com/roemhildtg/esri-promise/commit/970fcc092b6efe7cb1e5ef57d0c3feb11ee42db8). One of the benefits of esri-loader is that [you wouldn't need to do that](https://github.com/Esri/esri-loader#promises).

Anyway, I wanted to get your take on using esri-loader w/ StealJS, and let you know about esri-loader's new promise based API in case you wanted to use that in place of esri-promise. If you do and you want help w/ a PR, lmk.

**Comment by green3g** [comment]
@tomwayson that was very thorough! Thanks for reaching out, I've got a lot of comments.

> I was a fan of CanJS back in the day 

Isn't CanJS just a pleasure to work with? Its definitely not the most popular framework on the market, but I've been using it for a few different apps and decided to continue that trend with this project. 

> at one point you may have used esri-loader?

Yep, I started using the esri loader, with the goal of using native import statements for dojo modules, but I unfortunately wasn't able to get it to build correctly and then discovered esri-promise, which seemed like a more workable approach at the time. I'd definitely reconsider esri-loader. How I'm loading modules right now [feels very dirty](https://github.com/roemhildtg/can-arcgis/blob/master/components/draw-widget/ViewModel.js#L87) and I'd love to simply use `import Map from 'esri/map!dojo';` To truly integrate the esri api with steal though, I think the api would need to drop its use of dojo plugins like `dojo/i18n`, and its other conditional stuff like `dojo/has`. Alternatively we'd need a combination of steal plugins and extensions to handle these special cases. [steal-conditional](https://github.com/stealjs/steal-conditional/blob/master/conditional.js) does this. Steal itself has made a ton of progress since I last attempted it so it probably is worth revisiting. 

>  I'm guessing that b/c I see you exclude esri modules from the steal build but I don't see why you need that now.

Yeah, that particular line in the build isn't used anymore. There are a few reminiscent pieces of code floating around from things that I've played around with that I need to clean up. 

Regarding the optionsPromise comment, I'm not sure if that would work like you described, (I have yet to try it out but optionsPromise needs to be a `Promise` that resolves to the actual widget `options` in this case. I think we'd still need to wrap the `loadModules` in a `new Promise` and call `resolve`. Not sure though, I'd definitely be willing to try out a pr, and your help would absolutely be welcome on this project :)

**Comment by tomwayson** [comment]
> How I'm loading modules right now feels very dirty and I'd love to simply use import Map from 'esri/map!dojo';

Yeah, I know what you mean, but I've recently come to appreciate how explicit esri-loader/esri-promise are and really [think it's the best work around until we can actually load JSAPI modules directly](http://tomwayson.com/2018/01/05/loader-of-the-things-one-library-to-load-them-all/).

> I think the api would need to drop its use of dojo plugins like dojo/i18n, and its other conditional stuff like dojo/has.

Yep, they're working on that.

> Alternatively we'd need a combination of steal plugins and extensions to handle these special cases. steal-conditional does this.

Yeah, that could probably be done. There are similar efforts to create plugins for [webpack](https://github.com/OpenNTF/dojo-webpack-plugin), and [SystemJS](https://github.com/beginor/systemjs-plugin-dojo) that can load Dojo 1 modules, plugins like i18n and all.

That seems like a lot of effort though, and I worry that the build output might be larger than that of the CDN Dojo builds b/c those do [exclude certain modules that are loaded asynchronously as needed](https://github.com/Esri/arcgis-js-api/blob/6c4094654b085dc3b2736dc7cffc29ab47cf219e/esri.profile.js).

All in all my goal w/ esri-loader is to make sure there is _something_ that will work for all the module loaders out there until it's possible to load and bundle JSAPI modules directly. As dirty as the code may feel - it is roughly the equivalent of [dynamic `import()`](https://developers.google.com/web/updates/2017/11/dynamic-import) which is all the rage right now. Maybe thinking of it that way will help you sleep at night. 😉 

Tag me on a PR if you make one.

**Comment by green3g** [comment]
@tomwayson I created a new pull request with the esri-loader. You're right it works the same as my modified esri-promise native promise lib. 

I also added an example of using `optionsPromise`. https://github.com/roemhildtg/can-arcgis/pull/12/commits/863e3070610be98e8cdf0356755d0e17cc4c2602

I wrapped it in a new Promise to get it to work, but I think if I implemented perhaps a new method like `getOptions` with async/await, then we could use `return {options};` in that function...I'll play around with that a bit.
